### ISSUE TITLE
Add ability to configure ‘upload’ and ‘storage’ of videos separately when test suite passes/fails.

### ISSUE THREAD

**Issue by GrayedFox** [cypress-io/cypress#2522]
### Current behavior:

Currently, it's only possible to switch the video feature on or off in a binary way.

### Desired behavior:

To be able to record and process videos only when a test fails. If there's no way to do this (due to the recording feature not knowing whether a test will fail or not when it begins to record) - the processing part could at least be tweaked so that only failed tests have their videos processed and saved, whereas passing tests are technically recorded, but never processed or saved to disk if the test passes.

### Versions

Cypress 3.1.0
Chrome 69
Electron 59

**Comment by drewbrend** [comment]
@GrayedFox This already exists, but the docs are not clear.  https://github.com/cypress-io/cypress/issues/2217

**Comment by GrayedFox** [comment]
And for those who want to disable uploading entirely and still save failed spec videos?

I don't think it makes a lot of sense to bundle this into a single config switch - I can certainly see a use case where a user has a parallel setup doesn't want to upload any videos, but still wants failed videos to be saved on some machine somewhere. Uploading and recording are two very different things, and I'm now (at the very least) the second person to not understand that the `videoUploadOnPasses` partly covers this wanted behaviour.

Having two switches makes it super clear:

```
"uploadVideos": true | false
"recordVideos": true | false | fail | pass
```

Now, I can save only passing or failing videos, both, or none, and I will upload whatever videos happened to be there.

**Comment by drewbrend** [comment]
I agree that it leaves some functionality to be desired, but the current behaviour seems to do what you and I are asking for.

**Comment by GrayedFox** [comment]
I might take a look into it then - that is one of the points of open sourcing things I suppose! - if you have any hints for where to start looking, they would be very welcome :+1:

**Comment by drewbrend** [comment]
Check here:  https://github.com/cypress-io/cypress/blob/develop/packages/server/lib/modes/run.coffee#L716

**Comment by plemarquand** [comment]
@GrayedFox I like your proposal for `recordVideos`. Currently videos are generated for every suite on my build server but really I only want to keep the ones for failed suites as build artifacts, discarding the videos for passing suites to save space on the build server. 

As it stands right now setting `videoUploadOnPasses: false` is actually worse for my situation, since videos of passing suites are still recorded and saved to disk uncompressed.

**Comment by GrayedFox** [comment]
Ahoy, sorry I haven't forgotten about this, been pretty busy at work (and a while since I've touched coffee script!) - I think I'll be finished on this other side project soon and dedicate a weekend to this. Of course if someone else wants to jump in feel free 😄

**Comment by GrayedFox** [comment]
@brian-mann in response to your comment [here](https://github.com/cypress-io/cypress/issues/2652#issuecomment-443464329) - where you ask why this feature is necessary, I think this thread highlights valid use cases, perhaps?

Also I would argue that rather than asking the contributors and users of Cypress why they want to be able to control how they use Cypress, I would ask (aside from having the video recording have default values that you can build other features on top of it): is the video recording feature really so tightly coupled with dashboard functionality that it cannot be disabled without breaking the dashboard?

I'd need to take a better look at the code but that sounds like we may be breaking the single responsibility principle here. Ideally, the dashboard functions as a graphical interface for whatever pluggable features the Cypress user thinks they need... even if you know better and think that real life use cases would necessitate that all people should be using feature X or Y, that doesn't seem to me to be a reason to have such tightly coupled code. That, and using something like feature flags for the dash also means you can rapidly disable any feature which (god forbid!) ships with a serious bug, while working on a hotfix, without disabling other parts of the system that works (just food for thought!).

Thanks again for steering this wonderful product through the murky waters of open source shenanigans - I'm sure comments like these ones bring you immense happiness and joy 🙇

**Comment by kjellski** [comment]
I'm completely on your page @GrayedFox, very thankful for the cypress project itself and would not be testing my frontends like this if it wouldn't be there...

@brian-mann @jennifer-shehane let's focus on a solution here, the other ticket is closed and probably won't get much attention for others following along currently.

For reference, here are the last two comments on my deprecated issue, cited:

@brian-mann: 

> @kjellski I'm coming back to this now, and the answer may be in front of me - but I'm confused still as to why this is necessary. As of today, you don't need to record to the dashboard to get value out of the video. The option videoUploadOnPasses only affects when --record. If you're not using the Dashboard then you can manage the video yourself since the videos still be physically written to disk.

> I must be missing something. Your comment here: #2652 (comment) suggests that you're not using the Dashboard, and I just want to make sure we're on the same page that you can completely manage the video yourself if that's the case without us needing to change anything on our end.

@kjellski:

> Thanks for getting involved again @brian-mann! So the problem in general is this:
    AFAIK you can not use video recording without sending it out to your servers via upload.

> So you would be able to only record videos, without upload? I mean that's what I'm after... but it's not possible to use --record without a key to the dashboard and then it uploads... maybe I'm missing some documentation that suggested this... :D

**Comment by dialex** [comment]
Is there any update on this feature request?

**Comment by marinakuranda** [comment]
This feature would be very appreciated.
Ideally, it would follow the same logic as with saving screenshots - only saving them on fail.

**Comment by seyfer** [comment]
Here is another issue, not linked yet to this one, since it was closed.
https://github.com/cypress-io/cypress/issues/775

I really want it to happen. We do not need videos for each test on CI, we need video only for failing test, as it works with screenshots.

**Comment by Alados** [comment]
Any updates about the feature of storing videos (without using Dashboard) only for failed tests?

**Comment by kjellski** [comment]
@Alados @dialex @marinakuranda this is possible already, but I haven't looked into any updates on the discussions! The funny part is, it was possible from the beginning. You just don't specify the projectId part and set `video` to `true` as well as `videoUploadOnPasses` to `false`. We use the attached config for us just fine and the videos are just as if you'd run cypress locally. Then just mark them as artifacts in your CI and you have them available without using the cypress dashboard. 

I hope this helps :)

```json
{
  "baseUrl": "http://localhost:8080",
  "env": {
    "NODE_ENV": "local"
  },
  "reporter": "junit",
  "reporterOptions": {
    "mochaFile": "test-results/results.[hash].xml"
  },
  "video": true,
  "videoCompression": false,
  "videoUploadOnPasses": false,
  "viewportHeight": 900,
  "viewportWidth": 1440,
  "modifyObstructiveCode": false
}
```

**Comment by dialex** [comment]
Well @kjellski, my `cypress.json` file has that exact configuration (no `projectId`, `"video": true`, `"videoUploadOnPasses": false`) and still Cypress generates videos for all tests, which is not what we want.

To reiterate, we want the ability to record videos only for failed tests. A possible implementation was described at https://github.com/cypress-io/cypress/issues/2522#issuecomment-424358181:

```
"uploadVideos": true | false
"recordVideos": true | false | fail | pass
```

I think it's clear by now that the community wants this feature. The specification was already laid out by @GrayedFox. Is there anything blocking the implementation of this feature?

**Comment by kjellski** [comment]
@dialex aaahhhh, you only want the ones for failed! Sorry - I was not expecting this restriction - just as a workaround, you could just delete the successfull ones with a little script before ending your 'build' for now...

But yes, this seems to be not existing yet AFAIK!

**Comment by josefsmrz** [comment]
Any updates on this?

**Comment by GrayedFox** [comment]
> @dialex aaahhhh, you only want the ones for failed! ... you could just delete the successfull ones with a little script before ending your 'build' for now...

Ahhhh this old thread! Just to clarify - even if people only want to record videos for failed tests, it's impossible to know if a test will fail or not while the test is still running (since that would mean peering into the future, which not even Google's engineers can do - unless they know something I don't!) - which means recording videos only for failed tests in reality means recording videos for all tests and deleting/handling videos for tests which pass in some way. What @kjellski is suggesting, to me, makes sense.

It would indeed be possible to process a recording and not actually write anything to disk, given enough RAM, which could potentially be more efficient, depending on your/your cloud's setup. Disk space being cheap and SSD write speeds being so blazingly fast nowadays, recording and deleting videos would probably only cost a few milliseconds more per test (theory crafting now, don't quote me on this).

Anywho, TLDR: recording videos only for failed tests means recording videos for all tests, how one handles removing/deleting videos for passing tests is something that is either hidden from the user (Cypress handles it) or not 🤷‍♂️

**Comment by josefsmrz** [comment]
The point of being able to only save videos of the tests that failed is to avoid **processing** the videos captured for passing tests as that seems to be the most time-consuming operation in the process.

Don't get me wrong, I too understand that Cypress isn't designed to be a magic orb to see the future with and that I could just delete the videos of passing tests in an after-script, but that way I'm not able to benefit from the speed increase that I would get by not having to **process** them.

**Comment by jennifer-shehane** [comment]
Yes, how @GrayedFox explained is how this would need to be implemented. 

This issue is still in the 'proposal' stage, which means no work has been done on this issue as of today, so we do not have an estimate on when this will be delivered.

**Comment by mikaelboff** [comment]
Waiting for this!!!

**Comment by Jmcobb** [comment]
People at @shipt would love a feature like this.

**Comment by coltadorsey** [comment]
**Not** generating videos for skipped tests would be a welcomed consideration :)

**Comment by AskMarvinV** [comment]
Waiting for this as well.. :)

**Comment by rubiii** [comment]
We're also interested in this

**Comment by liambutler** [comment]
Also interested in this. I suspect that the saving of videos to Jenkins is responsible for our performance issues

**Comment by mikaelboff** [comment]
any updates on this?

**Comment by lucianojss** [comment]
I am also interested since at the moment we don't use cypress dashboard, and we keep videos in gitlab-ci artifacts and since our tests are starting to grow the size required for the artifacts also grow a lot and main interest is to keep only videos from the failed tests.

**Comment by davidmh** [comment]
> @lucianojss [...] keep only videos from the failed ones.

You can find the videos for failed tests using the screenshot directories since those get generated only for failing cases:

```bash
mkdir $path_to_your_artifacts
for scenario in $(ls ./cypress/screenshots); do
  cp "cypress/videos/$scenario.mp4" $path_to__your_atifacts
done
```

**Comment by rclmenezes** [comment]
@davidmh that's a brilliant hack. Nice

**Comment by Tomas-Tr** [comment]
somebody mastered to implement the hack from @davidmh into CI pipeline which uses copyfiles and publishartifacts task?

**Comment by jflayhart** [comment]
Hacks and other solutions like "just remove all the non-failed ones" is silly. This is not what we want! A simple toggle to NOT process SUCCESSFUL videos AT ALL. PERIOD. Regardless of using recording, cypress dashboards, etc.

The problem with said hacks is that it completely doesn't fix the problem that TIME and SPACE is required to store process, store, and generate these videos and that time is spent wasted for videos some users of the community don't need. We use Cypress liberally and the tests take minutes just to process videos we don't need.

For example, 99 tests pass, 1 test fails, but 100 videos are processed. In this example, we only want **ONE VIDEO RECORDING GENERATED AND SAVED.** I think there's still confusion around this ask.

**Comment by lizell** [comment]
This is mostly a performance issue. *Why spend time on encoding videos of passed tests, just to remove them later?* All that is needed is a config switch to tell Cypress that *we do not care about encoding/keeping the video of a passed test*. Much like screenshots already work.

Of course we need to capture the video for each test since we do not know in advance if it will fail or not, but we could skip the time consuming encoding part and saving of the video for a passed test.

**Comment by jennifer-shehane** [comment]
I've closed the issue about appending `(failed)` to the video filenames as a duplicate of this issue. #6998 The main motivator of that issue seemed to be what would be achieved in this issue - that people want non-passing tests' videos to be deleted.  

If you have a need for the video having a `(failed)` string appended to the video filename, we'd be open to hearing use cases for this outside of deleting the non-passing videos.

**Comment by jflayhart** [comment]
> The main motivator of that issue seemed to be what would be achieved in this issue - that people want non-passing tests' videos to be deleted.

Thanks for replying @jennifer-shehane I still think there's a misunderstanding here. It's not that "people want non-passing tests' videos to be deleted", that is in fact the inverse of precisely what we're asking 😄 We don't want failed tests to be removed, we want failed tests to be the _only_ things processed and saved, in other words, _passing_ tests should _not_ be saved, only failed ones.

**Comment by tom10271** [comment]
We are running Cypress with GitHub Action. We have limited number of tests to run only but total size of videos generated is not small. Downloading all of them to check what went wrong in a single test is not troublesome.

**Comment by tom10271** [comment]
> @davidmh that's a brilliant hack. Nice
> 
> Solution for recursive folders:
> 
> ```shell
> for s in $(find $SCREENSHOT_PATH -type f | sed "s|$SCREENSHOT_PATH\/||" | sed "s|\.png||"); do
>   mkdir -p $(dirname $TMP_PATH/$s) && cp "$VIDEO_PATH/$s.mp4" $TMP_PATH/$s.mp4
> done
> ```

Can you help to fix the bash script so that it supports filename with space?
I have tried to add ` | sed 's/ /\\ /g'` but not working. I am not familiar with bash script :(

**Comment by tom10271** [comment]
> > @davidmh that's a brilliant hack. Nice
> > Solution for recursive folders:
> > ```shell
> > for s in $(find $SCREENSHOT_PATH -type f | sed "s|$SCREENSHOT_PATH\/||" | sed "s|\.png||"); do
> >   mkdir -p $(dirname $TMP_PATH/$s) && cp "$VIDEO_PATH/$s.mp4" $TMP_PATH/$s.mp4
> > done
> > ```
> 
> Can you help fixing the bash script so that it supports filename with space?
> I have tried to add ` | sed 's/ /\\ /g'` but not working. I am not familiar with bash script :(

I have figured it out. I use TypeScript so it might not works the same as JS user does.

```bash
#!/usr/bin/env bash

IFS=

rm -rf ./test-artifacts
mkdir -p ./test-artifacts/videos

chmod -R 777 ./cypress/screenshots
cp -R ./cypress/screenshots ./test-artifacts/screenshots

find ./cypress/screenshots -iname *.ts | sed "s|./cypress/screenshots\/||" | sed "s|\.ts.*||" | while read -r s
do
    number_of_slash=$(echo -n ${s} | tr -cd '/' | expr $(wc -c) | sed 's/^ *//')

    if [ "$number_of_slash" -gt 0 ]; then
        mkdir -p $(dirname ./test-artifacts/videos/$s)
    fi

    cp "cypress/videos/$s.ts.mp4" "./test-artifacts/videos/$s.mp4"
done
```

**Comment by mikaelboff** [comment]
We want failed videos not to be processed and saved, any updates on this? @jennifer-shehane

**Comment by lizell** [comment]
> We want failed videos not to be processed and saved, any updates on this? @jennifer-shehane

No, we want videos for _passed_ tests not to be processed (or saved). At least for us a significant amount of time is being spent processing 100s of videos, that we never look at. Only the failed ones are interesting.

**Comment by mikaelboff** [comment]
> > We want failed videos not to be processed and saved, any updates on this? @jennifer-shehane
> 
> No, we want videos for _passed_ tests not to be processed (or saved). At least for us a significant amount of time is being spent processing 100s of videos, that we never look at. Only the failed ones are interesting.

Oops, sorry, I mistyped, just edited.

**Comment by dipasqualew** [comment]
I agree that we could save lots of time and CI resources by having the option to just skip processing for videos that we don't care about. Any update on this matter @jennifer-shehane?

Thanks!

**Comment by lizell** [comment]
Actually, some good news. I tried this maybe some year ago and at the time it didn't work (and the documentation states that it only works for the Dashboard). However I inspected the code and it turns out setting the config key `videoUploadOnPasses` to `true` does exactly what we want. It skips compression (most important) and upload (if Dashboard). I just verified it using Cypress 5.3.0.

So @jennifer-shehane, maybe we could just update the documentation (and/or optionally rename the config key)?

**Comment by khitrenovich** [comment]
> Actually, some good news. I tried this maybe some year ago and at the time it didn't work (and the documentation states that it only works for the Dashboard). However I inspected the code and it turns out setting the config key `videoUploadOnPasses` to `false` does exactly what we want. It skips compression (most important) and upload (if Dashboard). I just verified it using Cypress 5.3.0.
> 
> So @jennifer-shehane, maybe we could just update the documentation (and/or optionally rename the config key)?

I just tried that, and checked the results when one of the four spec files of a run failed.
Good news - the videos of the successful specs are not processed (as expected).
Bad news - if one of the specs fail, all the videos are uploaded as is, uncompressed, taking x10 more space.

So, using `videoUploadOnPasses` is not a real solution here.

**Comment by lizell** [comment]
OK, so for me I have a small script that removes the video files of passed tests before the upload stage. It is kind of a hack though, inspecting the resulting files of the `mocha-junit-reporter`.
In `cypress.json`set something like:
```
"reporter": "../../node_modules/mocha-junit-reporter",
"reporterOptions": {
  "mochaFile": "../../reports/junit/ui/[hash].xml"
},
```
and the my hack:
```bash
#!/usr/bin/env bash

echo Cleaning up videos:

for REPORT in reports/junit/ui/*.xml
do
  FILE=$(xmllint ${REPORT} --xpath '//@file')
  VIDEO=${FILE/ file=\"integration/tests\/ui\/videos}
  VIDEO=${VIDEO/.js\"/.js}.mp4

  if [ "$(xmllint ${REPORT} --xpath '//@failures>0')" = true ]; then
    echo " Keeping $VIDEO of a test with failure"
  else
    echo "Removing $VIDEO of a passed test"
    rm -f "$VIDEO"
  fi
done
```

You might want to adjust some paths above and there are certainly a number of other ways this could be achieved, but I leave this here as one way to do it ;).

**Comment by dipasqualew** [comment]
In all honesty I am starting to wonder the utility of the video feature. Nowadays we all have access to cheap, reproducible pipelines. Personally I run my E2E against a build I can spin up again, and if tests fail I can just point my local Cypress against the site.

However, this is not to say that I think the video feature is working "ok" as of today. While pipelines are cheap, time is not, and being forced to either process all videos, or to upload them, even when they are successful (if any of them fail), is very bad. It adds lots of time in the development-test cycle.

I know the options here are tailored for the Cypress Dashboard, but not everyone uploads these videos there. Major pipelines providers also allow artifact uploads.

**Comment by erickwilder** [comment]
> Actually, some good news. I tried this maybe some year ago and at the time it didn't work (and the documentation states that it only works for the Dashboard). However I inspected the code and it turns out setting the config key `videoUploadOnPasses` to `false` does exactly what we want. It skips compression (most important) and upload (if Dashboard). I just verified it using Cypress 5.3.0.
> 
> So @jennifer-shehane, maybe we could just update the documentation (and/or optionally rename the config key)?

This doesn't really solve the issue because files are still saved, which means that Cypress spends some resources recording these videos anyway. I would love to have a way (like the one describe by the @GrayedFox ) to choose which videos to record (and save to disk - `recordVideos`) and whether or not they should be uploaded to the Dashboard (`uploadVideos`).

Right now I use my own CI-related functions to save artifacts and expose videos internally to the team (not the Dashboard) if a CI pipeline fails. The problem is that even passed test suites includes tons of useless videos in the pipeline artifacts because we can't turn recording off.

**Comment by lizell** [comment]
But @GrayedFox is requesting exactly that – only failed videos are processed and saved. It is impossible to record only failed tests, since we can not know in advance if a test will fail or not. So what we want is;

* videos are recorded
* failed ones are saved and processed
* successful ones are not processed and discarded

`videoUploadOnPasses: false` takes us pretty close. But we still need native support to remove passing videos so that we don't have to write "hacks" to do it. We also have our own CI save artifact thingy, so we share the same problem. For us it is solved with the ["hack"](https://github.com/cypress-io/cypress/issues/2522#issuecomment-703483701).

**Comment by rooby** [comment]
> In all honesty I am starting to wonder the utility of the video feature. Nowadays we all have access to cheap, reproducible pipelines. Personally I run my E2E against a build I can spin up again, and if tests fail I can just point my local Cypress against the site.

It's definitely useful to save time in debugging tests running in CI pipelines.
Running in local cypress is not a viable solution, since sometimes failures occur specifically in the CI pipeline but not locally (for various potential reasons). Also, I don't want to have to run the tests again to investigate when I could instead just have artifacts from the original failed run.

In addition to the issue of time, disk space is unfortunately not unlimited. If your pipelines save their artifacts for a couple of weeks and you have a lot of pipelines running a lot of tests, keeping videos of all passing tests can waste a huge amount of drive space. Especially if you leave those passing test videos unprocessed.

**Comment by jennifer-shehane** [comment]
Sorry, that was a typo on my part. Everyone jumped on that. 😅 I clarified in my comment that people want passing videos deleted (non-failed videos). https://github.com/cypress-io/cypress/issues/2522#issuecomment-682365040

We do understand what people are asking - that this is an issue with performance, time and resources in CI, regardless of Dashboard use. Sorry for the confusion.

We're evaluating a possible API for more fine grained video configuration.

**Comment by jennifer-shehane** [comment]
Our team had some discussion around this with the following upcoming proposals:

### Video Compression

- Everyone can turn off video compression during `cypress run` today by setting [`"videoCompression": false`](https://on.cypress.io/configuration#Videos) in your Cypress configuration. This will save the time taken to compress ("process") the videos on every specfile that is run. We suggest setting this if you want to reduce the compression time spent in your runs.
- For people recording to the Dashboard, you can turn off video compression and uploading to the Dashboard on passing specs by setting [`"videoUploadOnPasses": false`](https://on.cypress.io/configuration#Videos) in your Cypress configuration today.
- For people recording to the Dashboard, we are discussing the possibility of handling video compression after uploading the video - so this will not take up time in the run on passing or failing videos. If delivered, the `videoUploadOnPasses` options would no longer be necessary.

### Deleting Videos

Instead of offering some form of 'deletePassingVideos' option to delete the videos if all the tests pass, we're proposing to expose an "after:spec" event that can be listened to in the plugins file where you can access the recorded video. This is where one could decide to delete the video if the spec has passed (or met any other defined requirement). 

*There are a few reasons we're proposing this way:*

1. We're planning to move away from our previous convention of releasing new configuration options for every use case requested. We think this is less flexible and we can meet more feature requests by offering interfaces to allow you to customize Cypress and your tests however you like. 
2. In situations where `retries` is turned on and you have a flaky test (2/3 attempts failed), where all tests are passing in the specfile, we believe the video is very valuable in this situation to evaluate why the flake happened. Offering a blanket 'deletePassingVideos' option will not give the flexibility to save videos for specfiles where flaky tests were present and all tests passed. 

The "after:spec" event listener would look similar to other event listeners available. See https://github.com/cypress-io/cypress/issues/6665

```js
module.exports = (on, config) => {
  // PROPOSAL: THIS DOES NOT CURRENTLY WORK
  on('after:spec', (arg1, arg2) => {
    // access video and delete if you want
  })
}
```

When this is released, we will provide the code necessary for you to write to delete videos if all tests in the spec passed.

### Feedback

We would love to hear any feedback on the proposals above. Feel free to comment in this issue with your thoughts.

**Comment by liambutler** [comment]
In my company, I'm currently working to help other teams to move on to Cypress. Our setup runs Cypress on Jenkins hosts, and are orchestrated with Cypress Dashboard. The videos are automatically saved in Jenkins. There is no way to get around this, other than completely disabling video in the configuration.

One of the current issues hindering wider adoption is how much space these videos take up per Jenkins Job, so disabling compression isn't an option for us.

While deleting the videos after the fact would help with this space issue, the videos will still be saved in the first instance. I'd like to increase the parallelisation on our cypress tests, but we are currently running into i/o bottlenecks on Jenkins. Saving then deleting in quick succession will not help with this

Deleting a saved video in an `after` step, while helpful for saving space, feels more like a bandaid to me. Given that there are already some video configuration options, I would rather that these be fleshed out.

**Comment by rooby** [comment]
@liambutler 

> While deleting the videos after the fact would help with this space issue, the videos will still be saved in the first instance.

aside from disabling videos entirely, how would you expect to solve your problem?
The video has to be created before knowing if the test will fail or not, so it has the be deleted after the fact.

The only thing I can think of that would reduce space, is to ensure that you can hook into the after spec as each spec is run and not only right at the end of all tests, so if all the tests pass the most space that would be used is the space of the largest single video file.

@jennifer-shehane 
The proposed solution seems acceptable,  although it would be a nice-to-have for videos and screenshots to have equivalent configuration options (not really necessary though, just seems more symmetrical 😆).
There would need to be documentation on how to implement the hook, since it would be a common requirement.

**Comment by liambutler** [comment]
@rooby 

> aside from disabling videos entirely, how would you expect to solve your problem?

Given that there is a saving/compression step after each spec, I am assuming that the videos are in memory until they are saved to disk. Is that not the case?

**Comment by GrayedFox** [comment]
@liambutler if you really want to only record videos for failed tests, the solution would be to run all failed tests twice (and hope you get the same failure again, which in my experience is likely, but this will cause headaches for flaky tests/envs).

You could set up a default config where videos are disabled entirely, and then you identify each failed test (easy enough to do after the fact or during runtime). Save the name of each failed test and construct a comma delimited string and then re-run only those failed tests with videos enabled, so with a `cypress run --spec "$commaDelimitedFailedTestsString" --record --tag "failed"` . Will cost a bit more time but if space and running tests in parallel are your main concerns, you might actually see a slight performance boost in the initial run since videos are skipped entirely during the first run, and I assume a healthy codebase on your end means you have a fail rate of less than, say, 10% 😉 

@jennifer-shehane out of interest when exactly does that after hook trigger?

1. before the video is saved to disk or encoded (it's just been recorded and exists in some raw format in memory)
2. after the video is written to disk, but before it is encoded
3. after the video is written to disk and after it is encoded

If it's either at step 1 or 2 I believe this proposal solves my (and most others) concerns. If it's at step 3, then it only helps solve the space issue, since it allows for a delete operation after each passing spec instead of at the end of the test suite.

**Comment by liambutler** [comment]
Still hoping that the solution I had in mind is possible, but thanks for all the tips! I love how helpful the Cypress community is toward each other

**Comment by jennifer-shehane** [comment]
Related PRs to this feature:

- `after:spec` event: https://github.com/cypress-io/cypress/pull/14178
- Properly handle deleted videos: https://github.com/cypress-io/cypress/pull/9595

**Comment by jennifer-shehane** [comment]
In Cypress 6.2.0, you can now listen to [`after:spec`](https://on.cypress.io/after-spec-api) events in the plugins file when setting the `experimentalRunEvents` configuration option to `true`. 

This means that you can delete the video after a spec has run for any reason - including if all of the tests passed. 

This alone will not skip the processing/compression of the video. If you want to skip the processing time for the video, you can set the `videoCompression` to `false`, so that the compression/processing time will be skipped. 

**To skip processing time for all videos:**

`cypress.json`

```json
{
  "videoCompression": false
}
```

**To delete a video if all tests pass:**

`cypress.json`

```json
{
  "experimentalRunEvents": true
}
```

`cypress/plugins/index.js

```js
// 'npm install -D del' - https://www.npmjs.com/package/del
const del = require('del')

module.exports = (on, config) => {
  on('after:spec', (spec, results) => {
    if (results.stats.failures === 0 && results.video) {
      // `del()` returns a promise, so it's important to return it to ensure
      // deleting the video is finished before moving on
      return del(result.video)
    }
  })
}
```

Let us know if you have any feedback on this experimental feature.

**Comment by lexsmil** [comment]
https://docs.cypress.io/api/plugins/after-spec-api.html#Delete-the-recorded-video-if-the-spec-passed

**Comment by GrayedFox** [comment]
Unless people have specific concerns with how this issue is resolved using the experimental feature, I believe the core concerns have been addressed. Thank you to @jennifer-shehane for your continued patience and persistence in getting this resolved and navigating both the external and internal communications channels, and to all those who used this thread constructively.

As with many a good thing, they can and do take time.

**Comment by nichtmeinweg** [comment]
tried to use the approved solution but getting "fast-glob\out\readers\sync.js: Unexpected token"  on "cypress": "^6.2.1". Anything I miss? 

> In Cypress 6.2.0, you can now listen to [`after:spec`](https://on.cypress.io/after-spec-api) events in the plugins file when setting the `experimentalRunEvents` configuration option to `true`.
> 
> This means that you can delete the video after a spec has run for any reason - including if all of the tests passed.
> 
> This alone will not skip the processing/compression of the video. If you want to skip the processing time for the video, you can set the `videoCompression` to `false`, so that the compression/processing time will be skipped.
> 
> **To skip processing time for all videos:**
> 
> `cypress.json`
> 
> ```json
> {
>   "videoCompression": false
> }
> ```
> 
> **To delete a video if all tests pass:**
> 
> `cypress.json`
> 
> ```json
> {
>   "experimentalRunEvents": true
> }
> ```
> 
> `cypress/plugins/index.js
> 
> ```js
> // 'npm install -D del' - https://www.npmjs.com/package/del
> const del = require('del')
> 
> module.exports = (on, config) => {
>   on('after:spec', (spec, results) => {
>     if (results.stats.failures === 0 && results.video) {
>       // `del()` returns a promise, so it's important to return it to ensure
>       // deleting the video is finished before moving on
>       return del(result.video)
>     }
>   })
> }
> ```
> 
> Let us know if you have any feedback on this experimental feature.

**Comment by jennifer-shehane** [comment]
@nichtmeinweg I did mispell `results` in the original workaround. I've updated my comment, but this should not have given this error. Can you provide a reproducible example?

**Comment by nichtmeinweg** [comment]
> @nichtmeinweg I did mispell `results` in the original workaround. I've updated my comment, but this should not have given this error. Can you provide a reproducible example?

Thanks a lot for the correction, my  error is most probably related to the needed upgrade of cypress I had to do for following the workaround. 

`
[2]
[2] Oops...we found an error preparing this test file:
[2]
[2]   **cypress\support\index.js**
[2]
[2] The error was:
[2]
[2] Error: Parsing file C:\repo\patientstrength_KOS-447\ps-client-ekogu\ClientApp\node_modules\fast-glob\out\readers\stream.js: Unexpected token (16:10)
`

**Which version of "del" are you using?** 


it is not related to the flag: "experimentalRunEvents", or the EVENT function, its actually this line that causes the error: 
`const del = require('del');`




Dependencies:
`
 "dependencies": {
    "@coreui/coreui": "^2.1.16",
    "@coreui/react": "^2.5.8",
    "@date-io/date-fns": "^1.3.13",
    "@fortawesome/fontawesome-svg-core": "^1.2.28",
    "@fortawesome/free-solid-svg-icons": "^5.13.0",
    "@fortawesome/react-fontawesome": "^0.1.9",
    "@material-ui/core": "^4.10.2",
    "@material-ui/icons": "^4.9.1",
    "@material-ui/lab": "^4.0.0-alpha.56",
    "@material-ui/pickers": "^3.2.10",
    "@microsoft/applicationinsights-react-js": "^3.0.1",
    "@microsoft/applicationinsights-web": "^2.5.6",
    "@testing-library/jest-dom": "^5.11.5",
    "@testing-library/react": "^11.1.0",
    "@testing-library/user-event": "^12.1.10",
    "bootstrap": "^4.4.1",
    "chart.js": "^2.9.3",
    "classnames": "^2.2.6",
    "core-js": "^3.6.5",
    "cross-env": "^7.0.2",
    "cypress-commands": "^1.1.0",
    "cypress-file-upload": "^4.1.1",
    "date-fns": "^2.12.0",
    "enzyme": "^3.11.0",
    "enzyme-adapter-react-16": "^1.15.2",
    "history": "^5.0.0",
    "html-to-react": "^1.4.3",
    "i18next": "^19.4.4",
    "i18next-http-backend": "^1.0.8",
    "jquery": "^3.5.1",
    "jwt-decode": "^2.2.0",
    "lodash": "4.17.20",
    "mailslurp-client": "^10.0.4",
    "msal": "^1.4.3",
    "node-sass": "^4.14.1",
    "npm": "^6.14.7",
    "prop-types": "^15.7.2",
    "react": "^17.0.1",
    "react-aad-msal": "^2.3.5",
    "react-app-polyfill": "^1.0.6",
    "react-bootstrap": "^1.0.1",
    "react-dom": "^17.0.1",
    "react-dropzone": "^11.0.2",
    "react-i18next": "^11.4.0",
    "react-icd10": "^1.0.7",
    "react-icons": "^3.10.0",
    "react-js-banner": "^0.5.2",
    "react-keys": "^3.0.0-rc20",
    "react-live-clock": "^4.0.5",
    "react-overlay-component": "^1.0.2",
    "react-phone-input-2": "^2.13.3",
    "react-phone-number-input": "^3.0.22",
    "react-redux": "^7.2.0",
    "react-router-dom": "^5.1.2",
    "react-scripts": "^4.0.0",
    "react-scroll": "^1.8.0",
    "react-stepper-horizontal": "^1.0.11",
    "react-textarea-autosize": "^8.0.1",
    "react-window": "^1.8.5",
    "reactstrap": "^8.4.1",
    "redux": "^4.0.5",
    "redux-form": "^8.3.5",
    "redux-logger": "^3.0.6",
    "redux-thunk": "^2.3.0",
    "web-vitals": "^0.2.4"
  },
 "devDependencies": {
    "@babel/core": "7.11.4",
    "@babel/plugin-proposal-class-properties": "7.12.1",
    "@bahmutov/cy-api": "1.4.2",
    "@cypress/code-coverage": "^3.9.1",
    "@cypress/instrument-cra": "^1.4.0",
    "@cypress/webpack-preprocessor": "^5.5.0",
    "@storybook/addon-actions": "6.0.22",
    "@storybook/addon-essentials": "6.0.22",
    "@storybook/addon-links": "6.0.22",
    "@storybook/node-logger": "6.0.22",
    "@storybook/preset-create-react-app": "3.1.4",
    "@storybook/react": "6.0.22",
    "@types/jest": "26.0.14",
    "@types/node": "14.11.2",
    "@types/react": "16.9.54",
    "@types/react-dom": "16.9.9",
    "babel-loader": "8.1.0",
    "babel-plugin-istanbul": "6.0.0",
    "chalk": "4.1.0",
    "check-code-coverage": "1.9.3",
    "console-log-div": "0.6.3",
    "cypress": "^6.2.1",
    "cypress-xpath": "^1.6.1",
    "del": "^6.0.0",
    "electron": "10.1.5",
    "eslint": "7.14.0",
    "eslint-config-google": "0.14.0",
    "eslint-plugin-only-warn": "1.0.2",
    "eslint-plugin-react": "7.21.5",
    "express": "4.17.1",
    "fluent-ffmpeg": "^2.1.2",
    "istanbul-merge": "1.1.1",
    "jest-junit": "11.1.0",
    "lint-staged": "10.5.2",
    "markdown-link-check": "3.8.1",
    "mocha-junit-reporter": "2.0.0",
    "nyc": "15.1.0",
    "parcel-bundler": "1.12.4",
    "prettier": "1.19.1",
    "semantic-release": "^17.3.1",
    "serve": "11.3.2",
    "start-server-and-test": "^1.11.7",
    "storybook-addon-jsx": "7.3.4",
    "typescript": "3.8.3",
    "yarn": "1.22.7"
  }
`

**Comment by caseyjhol** [comment]
@jennifer-shehane 

> This alone will not skip the processing/compression of the video. If you want to skip the processing time for the video, you can set the `videoCompression` to `false`, so that the compression/processing time will be skipped.

Doesn't this disable compression for all videos then? I'd still like to compress videos for failed tests.

**Comment by josefsmrz** [comment]
> @jennifer-shehane
> 
> > This alone will not skip the processing/compression of the video. If you want to skip the processing time for the video, you can set the `videoCompression` to `false`, so that the compression/processing time will be skipped.
> 
> Doesn't this disable compression for all videos then? I'd still like to compress videos for failed tests.

It does. You can set `  "videoUploadOnPasses": false ` to disable compression only for passing tests.

**Comment by caseyjhol** [comment]
@josefsmrz That's only for Cypress videos uploaded to the Dashboard, though. It seems like deleting the video via the `after:spec` event prevents processing by default (#9595), so I think `videoCompression: false` can be avoided (but perhaps I'm wrong).

**Comment by josefsmrz** [comment]
> @josefsmrz That's only for Cypress videos uploaded to the Dashboard, though. It seems like deleting the video via the `after:spec` event prevents processing by default (#9595), so I think `videoCompression: false` can be avoided (but perhaps I'm wrong).

@caseyjhol I guess it is intended mainly for use with the Dashboard, but as a side effect, the setting actually does what I've described even when not using the Dashboard. It has been discussed [in the very beginning of this thread](https://github.com/cypress-io/cypress/issues/2522#issuecomment-424358181)

**Comment by kaiyoma** [comment]
> Let us know if you have any feedback on this experimental feature.

@jennifer-shehane This seems to work very well for us. We were running out of space on our CI server and large videos were the main culprit. In addition to freeing up lots of space, we save valuable time by not processing the video too. The rare fix that saves both time and space! 😄

**Comment by kaiyoma** [comment]
Is anyone else noticing (after implementing the suggestion here) that now there are warnings about missing video files? I'm seeing this for all my tests now:

```
18:53:39  Warning: We could not find the video at the following path, so we were unable to process it.
18:53:39  
18:53:39  Video path: .../basic/data.spec.js.mp4
18:53:39  
18:53:39  This error will not alter the exit code.
```

Is there any way to silence these warnings?

**Comment by josefsmrz** [comment]
> Is anyone else noticing (after implementing the suggestion here) that now there are warnings about missing video files? I'm seeing this for all my tests now:
> 
> ```
> 18:53:39  Warning: We could not find the video at the following path, so we were unable to process it.
> 18:53:39  
> 18:53:39  Video path: .../basic/data.spec.js.mp4
> 18:53:39  
> 18:53:39  This error will not alter the exit code.
> ```
> 
> Is there any way to silence these warnings?

yeah also got them

**Comment by jnv** [comment]
I get the same error about missing videos. Perhaps one possibility to avoid this is to collect file paths and delete videos in `after:run` event handler. I have adjusted the original code to something like:

```js
import {rm} from 'fs/promises'

const pluginConfig = (on, config) => {
	const filesToDelete = []
	on('after:spec', (spec, results) => {
		if (results.stats.failures === 0 && results.video) {
			filesToDelete.push(results.video)
		}
	})
	on('after:run', async () => {
		console.log(
			'Deleting %d videos from successful specs',
			filesToDelete.length
		)
		await Promise.all(filesToDelete.map((videoFile) => rm(videoFile)))
	})
}

export default pluginConfig
```

Not that this will delete videos only when ran through `cypress run`. Also unless you set either `videoCompression`, or `videoUploadOnPasses` options to `false`, the videos will be compressed only to be deleted which is wasteful.

Also, this particular snippet won't work with Node 12 currently bundled by Cypress due to lack of `fs/promises` module. I will leave this part as an exercise to reader. 😉

**Comment by NickRumenov** [comment]
Yes, that works without the error message, I just changed the "after:run" hook with:
```
on('after:run', async () => {

    filesToDelete.forEach(path => {
      fs.unlink(path, (err) => {
        if (err) {
          console.error(err)
          return
        }
        console.log('Successfully deleted passing test videos' + path)
      })
    })
  })
```

NodeJS 14.5.1 
Hopefully, we will have an official solution soon. Thanks

**Comment by edaldridge** [comment]
I've added this to my current setup, minus the experimental features part and the video compression which I believe are not needed for this to work fully now.

The process works locally for me but when I pushed the new plugin/index.js to my repo and ran a test from Jenkins I am getting an error about the plugin/index.js file not being found (see below).  I am guessing this is more than likely a config error on Jenkins rather than something related to the improvement suggested here but I am hoping someone else may have experienced this issue and may have a solution?

The plugins file is missing or invalid.

Your `pluginsFile` is set to `/opt/jenkins/workspace/<path to test>/cypress/plugins/index.js`, but either the file is missing, it contains a syntax error, or threw an error when required. The `pluginsFile` must be a `.js`, `.ts`, or `.coffee` file.

Or you might have renamed the extension of your `pluginsFile`. If that's the case, restart the test runner.

Please fix this, or set `pluginsFile` to `false` if a plugins file is not necessary for your project.

 Error: Cannot find module 'del'

**Comment by jennifer-shehane** [comment]
@kaiyoma We've updated the 'warning' message for when the video is missing for processing. Since this is likely expected behavior that someone coded in (to delete the video), we replaced the messaging with a single line so that you are still informed that the video processing is being skipped. https://github.com/cypress-io/cypress/pull/15828

```
No video found after spec ran - skipping processing. Video path: /my/videos
```

**Comment by jennifer-shehane** [comment]
The `experimentalRunEvents` configuration is no longer necessary to get this 'after:spec' listener. It is built into the product directly. 

This issue will be closed to further comment as the exact issue feature here is delivered. See https://github.com/cypress-io/cypress/issues/2522#issuecomment-749316813 for an example. 

If you're experiencing a bug related to the issue above in Cypress, please [open a new issue](https://github.com/cypress-io/cypress/issues/new/choose) with a fully reproducible example that we can run. There may be a specific edge case with the issue that we need more detail to fix.

**Comment by jennifer-shehane** [comment]
Uploaded this comment to show solution for controlling which videos to delete in newer versions: https://docs.cypress.io/guides/guides/screenshots-and-videos#Control-which-videos-to-keep-and-upload-to-Cypress-Cloud
